File: arrays/05-Maximum-subarray/Solution.java
package com.sandbox;

public class Solution {
  public int largestSubArraySum(final int[] numbers) {
    assert (numbers.length > 0);
    if (numbers.length == 1) return numbers[0];
    
    int start = 0;
    int end = numbers.length - 1;
    int max = 0;
    int sum = 0;
    
    while (start != numbers.length - 1) {
      sum = sumSubArray(numbers, start, end);
      if (sum > max) {
        max = sum;
      }
      
      end--;
      if (end == start) {
        start++;
        end = numbers.length - 1;
      }
    }
    
    return max;
  }
  
  private int sumSubArray(final int[] numbers, int startIndex, int endIndex) {
    int sum = 0;
    for (int i = startIndex; i <= endIndex; i++) {
      sum += numbers[i];
    }
    
    return sum;
  }
}

File: arrays/05-Maximum-subarray/solution.go
package main

import (
	"fmt"
)

func sumSubarray(nums []int, startIdx int, endIdx int) int {
	sum := 0
	for i := startIdx; i <= endIdx; i++ {
		sum += nums[i]
	}
	return sum
}

func MaximumSubarray(nums []int) int {
	max := len(nums)
	if max == 1 {
		return nums[0]
	}

	start := 0
	end := max - 1
	result := 0
	current := 0

	for start < end {
		current = sumSubarray(nums, start, end)
		if current > result {
			result = current
		}

		end--
		if end == start {
			start++
			end = max - 1
		}

		if start == max-1 {
			break
		}
	}

	return result
}

func main() {
	input := []int{-2, 1, -3, 4, -1, 2, 1, -5, 4}
	result := MaximumSubarray(input)
	fmt.Println(result)
}
